feat: add operation messages via `RAISE NOTICE`

benjie opened the pull request:
Fixes #12 (cc @demianuco)

PostGraphile will subscribe to `NOTICE` events from PostgreSQL during mutations, and any notice with error code `OPMSG` (operation message) will be added to the mutation payload `messages` field.

### Important:

- This feature only applies to mutations
- Any PL/pgSQL function may raise these notices
- Additional information can be added to the OPMSG by passing a JSON value to the `detail` field
- `RAISE NOTICE` may have message length limits; it's your responsibility to colour inside the lines
- Clients must be configured (via `client_min_messages`) to receive NOTICE messages; this is PostgreSQL's default configuration.
- If the server is configured (via `log_min_messages`) to log NOTICEs then, well, the NOTICEs will be logged; PostgreSQL's default configuration does not log NOTICE.

### Example:

```sql
raise notice 
  '2 + 2 is %, minus 1 that''s %; quick maths.',
  (2 + 2),
  (2 + 2 - 1)
using
  errcode = 'OPMSG',
  detail = json_build_object(
    'level', 'info',
    'path', array_to_json(ARRAY['noticePath']),
    'anything_else', 'can_go_here'
  )::text;
```

benjie committed:
First attempt at notice plugin

benjie committed:
Convert functions to plpgsql and add 'raise debug'; tests still pass

benjie committed:
Parse GOPHK NOTICEs

benjie committed:
Change code to OPMSG

benjie committed:
Lint fixes

benjie commented:
Rebased on #17 (after merge to master) to make the diff smaller and more focussed.

mathroc commented:
I don’t understand how notices are filtered, it looks like code different from `OPMSG` would be processed too, am I missing something ?

benjie committed:
Only capture OPMSG NOTICEs

benjie commented:
:eyes: I was so tired that day. I wasn't confident merging because I knew I'd overlooked something but couldn't figure out what. Thanks for spotting @mathroc! Added test and fixed.

demianuco commented on src/PgNoticeMessagesPlugin.ts:
```diff
@@ -0,0 +1,84 @@
+        context,
+        resolveInfo: GraphQLResolveInfoWithMessages
+      ) => {
+        assert(!processNotice, "processNotice should not be set yet!");
```
This assertion passes on the first request, but fails on subsequent ones. Commenting it out does not seem to have any ill effects, and makes everything work as expected.

benjie replied:
Fixed :+1:

benjie committed:
Correctly register and clear processNotice

benjie committed:
Final tweaks

benjie committed:
Update README
